File: src/mffs/item/module/ItemModule.java
package mffs.item.module;

import java.util.List;
import java.util.Set;

import mffs.api.IFieldInteraction;
import mffs.api.IProjector;
import mffs.api.modules.IModule;
import mffs.base.ItemMFFS;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.world.World;
import universalelectricity.api.vector.Vector3;

public class ItemModule extends ItemMFFS implements IModule
{
	private float fortronCost = 0.5f;

	public ItemModule(int id, String name)
	{
		super(id, name);
	}

	@Override
	public void addInformation(ItemStack itemStack, EntityPlayer player, List info, boolean b)
	{
		info.add("Fortron: " + this.getFortronCost(1));
		super.addInformation(itemStack, player, info, b);
	}

	@Override
	public Set<Vector3> onPreCalculate(IFieldInteraction projector, Set<Vector3> position)
	{
		return position;
	}

	@Override
	public void onCalculate(IFieldInteraction projector, Set<Vector3> position)
	{
	}

	@Override
	public boolean onProject(IProjector projector, Set<Vector3> fields)
	{
		return false;
	}

	@Override
	public int onProject(IProjector projector, Vector3 position)
	{
		return 0;
	}

	@Override
	public boolean onCollideWithForceField(World world, int x, int y, int z, Entity entity, ItemStack moduleStack)
	{
		return false;
	}

	public ItemModule setCost(float cost)
	{
		this.fortronCost = cost;
		return this;
	}

	@Override
	public ItemModule setMaxStackSize(int par1)
	{
		super.setMaxStackSize(par1);
		return this;
	}

	@Override
	public float getFortronCost(float amplifier)
	{
		return this.fortronCost;
	}

	@Override
	public boolean onDestroy(IProjector projector, Set<Vector3> field)
	{
		return false;
	}

	@Override
	public boolean requireTicks(ItemStack moduleStack)
	{
		return false;
	}
}

File: src/main/scala/mffs/base/ItemModule.scala
package mffs.base

import java.util.{List => JList, Optional, Set => JSet}

import com.resonant.core.prefab.itemblock.TooltipItem
import com.resonant.core.prefab.modcontent.AutoItemTexture
import com.resonant.wrapper.lib.utility.science.UnitDisplay
import mffs.api.machine.Projector
import mffs.api.modules.Module
import mffs.field.BlockProjector
import nova.core.entity.Entity
import nova.core.game.Game
import nova.core.item.Item
import nova.core.player.Player
import nova.core.util.transform.{Cuboid, Vector3i}

import scala.collection.convert.wrapAll._

abstract class ItemModule extends Item with TooltipItem with Module with CategoryMFFS with AutoItemTexture {
	private var fortronCost = 0.5f
	private var maxCount = 64

	override def getTooltips(player: Optional[Player], tooltips: JList[String]) {
		tooltips.add(Game.instance.languageManager.translate("info.item.fortron") + " " + new UnitDisplay(UnitDisplay.Unit.LITER, getFortronCost(1) * 20) + "/s")
	}

	override def getFortronCost(amplifier: Float) = fortronCost

	def setCost(cost: Float): this.type = {
		this.fortronCost = cost
		return this
	}

	override def getMaxCount: Int = maxCount

	def setMaxCount(maxCount: Int): ItemModule = {
		this.maxCount = maxCount
		return this
	}

	def getEntitiesInField(projector: Projector): Set[Entity] = {
		val blockProjector = projector.asInstanceOf[BlockProjector]
		val bound = new Cuboid(-projector.getNegativeScale, projector.getPositiveScale + Vector3i.one) + blockProjector.position() + projector.getTranslation

		return blockProjector.world.getEntities(bound)
			.filter(entity => projector.isInField(entity.position()))
			.toSet
	}
}